internal/modgen/modgen.go

```diff
@@ -0,0 +1,420 @@
+package modgen
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	"math"
+	"math/rand"
+	"strconv"
+
+	"github.com/tetratelabs/wazero/internal/leb128"
+	"github.com/tetratelabs/wazero/internal/wasm"
+)
+
+// Gen generates a pseudo random compilable module based on `seed`. The size of
+// the generated module corresponds to the size of the seed. For example,
+// the more larger len(seed) is, the more functions the module *likely* has.
+//
+// TODO: make it possible to explicitly control the size of each section.
+// TODO: have an option to have whether or not the result has imports.
+//
+// Note: "pseudo" here means the determinism of the generated results,
+// e.g. giving same seed returns exactly the same module for
+// the same code base in Gen.
+//
+// Note: this is only used for testing wazero runtime.
+func Gen(seed []byte) *wasm.Module {
+	if len(seed) == 0 {
+		return &wasm.Module{}
+	}
+
+	checksum := sha256.Sum256(seed)
+	// Use 4 randoms created from the unique sha256 hash value of the seed.
+	g := &generator{size: len(seed), rands: make([]random, 4)}
+	for i := 0; i < 4; i++ {
+		g.rands[i] = rand.New(rand.NewSource(
+			int64(binary.LittleEndian.Uint64(checksum[i*8 : (i+1)*8]))))
+	}
+	return g.gen()
+}
+
+type generator struct {
+	// rands holds random sources for generating a module.
+	rands         []random
+	nextRandIndex int
+
+	// size holds the original size of the seed.
+	size int
+
+	// m is the resulting module.
+	m *wasm.Module
+}
+
+// random is the interface over methods of rand.Rand which are used by our generator.
+// This is only for testing the generator implmenetation.
+
+type random interface {
+	// See rand.Intn.
+	Intn(n int) int
+
+	// See rand.Read
+	Read(p []byte) (n int, err error)
+}
+
+func (g *generator) nextRandom() (ret random) {
+	ret = g.rands[g.nextRandIndex]
+	g.nextRandIndex = (g.nextRandIndex + 1) % len(g.rands)
+	return
+}
+
+func (g *generator) gen() *wasm.Module {
```

Comment:
docs can say punt to individual sections for details on what's generated.

each should summarize what's likely to be here. Ex will there always be a start function? will there always be memory. will there be a name section?